packages/shared/src/components/buttons/ButtonV2.tsx

```diff
@@ -0,0 +1,150 @@
+};
+
+export type IconType = React.ReactElement<IconProps>;
+
+export { ButtonColor };
+
+export enum ButtonKind {
+  Button = 'button',
+  Link = 'link',
+}
+
+export enum ButtonVariant {
+  Primary = 'primary',
+  Secondary = 'secondary',
+  Tertiary = 'tertiary',
+}
+
+export enum ButtonIconPosition {
+  Left = 'left',
+  Right = 'right',
+}
+interface CommonButtonProps {
+  kind?: ButtonKind;
+  size?: ButtonSize;
+  loading?: boolean;
+  pressed?: boolean;
+  disabled?: boolean;
+  children?: ReactNode;
+}
+
+// when color is present, variant is required
+type ColorButtonProps =
+  | { color: ButtonColor; variant: ButtonVariant }
+  | { color?: never; variant?: ButtonVariant };
+
+// when iconPosition is present, icon is required
+type IconButtonProps =
+  | { iconPosition: ButtonIconPosition; icon: IconType }
+  | { iconPosition?: never; icon?: IconType };
+
+type BaseButtonProps = CommonButtonProps & ColorButtonProps & IconButtonProps;
+
+const useGetIconWithSize = (size: ButtonSize, iconOnly: boolean) => {
+  return (icon: React.ReactElement<IconProps>) =>
+    React.cloneElement(icon, {
+      size: icon.props?.size ?? buttonSizeToIconSize[size],
+      className: classNames(icon.props.className, !iconOnly && 'icon'),
+    });
+};
+
+export type AllowedTags = keyof Pick<JSX.IntrinsicElements, 'a' | 'button'>;
+export type AllowedElements = HTMLButtonElement | HTMLAnchorElement;
+export type ButtonElementType<Tag extends AllowedTags> = Tag extends 'a'
+  ? HTMLAnchorElement
+  : HTMLButtonElement;
+
+export type ButtonProps<Tag extends AllowedTags> = BaseButtonProps &
+  HTMLAttributes<Tag> &
+  JSX.IntrinsicElements[Tag] & {
+    ref?: Ref<ButtonElementType<Tag>>;
+  };
+
+function ButtonComponent<T extends AllowedTags>(
+  {
+    kind = ButtonKind.Button,
+    variant,
+    size = ButtonSize.Medium,
+    color,
+    className,
+    icon,
+    iconPosition = ButtonIconPosition.Left,
+    loading,
+    pressed,
+    children,
+    onClick,
+    ...props
+  }: ButtonProps<T>,
+  ref?: Ref<ButtonElementType<T>>,
+): ReactElement {
+  const iconOnly = icon && !children;
+  const getIconWithSize = useGetIconWithSize(size, iconOnly);
+  const isAnchor = kind === ButtonKind.Link;
+  const Tag = isAnchor ? 'a' : 'button';
+
+  return (
+    <Tag
+      {...props}
+      {...(isAnchor ? combinedClicks(onClick) : { onClick })}
+      aria-busy={loading}
+      aria-pressed={pressed}
+      ref={ref}
+      className={classNames(
+        { iconOnly },
+        size,
+        'btn flex flex-row items-center justify-center border typo-callout font-bold no-underline shadow-none cursor-pointer select-none focus-outline relative',
+        {
+          [`btn-${variant}`]: variant,
+          [`btn-${variant}-${color}`]: color && variant,
+        },
+        className,
+      )}
+    >
+      {icon &&
+        iconPosition === ButtonIconPosition.Left &&
+        getIconWithSize(icon)}
+      {children && <span>{children}</span>}
```

Comment:
I thought we were supposed to not use the `span` from within the Button component reading the DR.

Reply:
There are some style definitions in button.css that rely on the button component to have a child component. Otherwise the box shadow was not applied correctly for example. There is another solution for this of course, but this seemed like the least amount of change.

The DR itself is talking about removing the `spanClassName` prop, not about removing the span itself.